競技プロ就活部PR用/102. Binary Tree Level Order Traversal.md

```diff
@@ -0,0 +1,88 @@
+15分で1回目は解けず、どのようにして同じ階層のノードを取り出して処理し、次の階層に移動するかが整理できませんでした。
+各階層から値を列挙した後に、その回数分for文を回すのは、BFSの基本ができていないから思いつかなかった(考えられなかった)のかなと思います。
+
+### 1回目
+時間計算量: O(N)
+空間計算量: O(N + 正解分)
+```python
+class Solution:
+    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
+        if not root:
+            return []
+
+        result = []
+        next_level_nodes = deque()
+        next_level_nodes.append(root)
+
+        while next_level_nodes:
+            same_level_nodes = []
+            for i in range(len(next_level_nodes)):
+                node = next_level_nodes.popleft()
```

Comment:
`popleft()` すると n = len(next_level_nodes) として、時間計算量 O(n) かかってしまいます。理由は、 popleft() した場合、右側の要素を全て 1 つずつ左にずらさなければならないためです。 (CPython の実装がそのようになっているかは確認しておりません。念のため確認をお願いいたします。)

node の list を 2 つ持ち、片方を for 文で回し、もう片方に子ノードを入れていき、 1 レベル終わったら後者を全車に代入する、という実装はいかがでしょうか？

Reply:
dequeなのでpopleftの計算量はO(1)ではないですか？

https://docs.python.org/3/library/collections.html#collections.deque

> Deques are a generalization of stacks and queues (the name is pronounced “deck” and is short for “double-ended queue”). Deques support thread-safe, memory efficient appends and pops from either side of the deque with approximately the same O(1) performance in either direction.

Reply:
list を使っていると見間違えました。失礼いたしました。

Reply:
nodhipさん、ahayashiさん
ありがとうございます。listを使って、popleftなしでも書き直してみます。

実装を確認してみました。
https://github.com/python/cpython/blob/main/Modules/_collectionsmodule.c

```c
static PyObject *
deque_popleft_impl(dequeobject *deque)
/*[clinic end generated code: output=62b154897097ff68 input=1571ce88fe3053de]*/
{
    PyObject *item;
    block *prevblock;

    if (Py_SIZE(deque) == 0) {
        PyErr_SetString(PyExc_IndexError, "pop from an empty deque");
        return NULL;
    }
    assert(deque->leftblock != NULL);
    item = deque->leftblock->data[deque->leftindex];
    deque->leftindex++;
    Py_SET_SIZE(deque, Py_SIZE(deque) - 1);
    deque->state++;

    if (deque->leftindex == BLOCKLEN) {
        if (Py_SIZE(deque)) {
            assert(deque->leftblock != deque->rightblock);
            prevblock = deque->leftblock->rightlink;
            freeblock(deque, deque->leftblock);
            CHECK_NOT_END(prevblock);
            MARK_END(prevblock->leftlink);
            deque->leftblock = prevblock;
            deque->leftindex = 0;
        } else {
            assert(deque->leftblock == deque->rightblock);
            assert(deque->leftindex == deque->rightindex+1);
            /* re-center instead of freeing a block */
            deque->leftindex = CENTER + 1;
            deque->rightindex = CENTER;
        }
    }
```
サイズやindexのインクリメントで、O(1)と思います。もし、参照先や読み方間違っていたらぜひ教えて欲しいです。

Reply:
申し訳ありません。 list で popleft() をしているのだと勘違いしました。失礼いたしました。

Reply:
いえ、アドバイスと確認の機会、ありがとうございました。
参考になりました。